pisek/visualize.py

```diff
@@ -13,304 +9,330 @@
+
+test_name_max_length = 0
+
+
+class MissingSolution(Exception):
+    def __init__(self, name: str) -> None:
+        super().__init__(f"Missing solution {name} in testing log.")
+
+
+@dataclass
+class LoggedResult:
+    verdict: Verdict
+    points: float
+    time: float
+    test: str
+    original_verdict: Verdict
+
+    def __init__(
+        self,
+        verdict: Verdict,
+        points: float,
+        measured_value: float,
+        test: str,
+        original_verdict: Verdict,
+    ) -> None:
+        global test_name_max_length
+
+        self.verdict = verdict
+        self.points = points
+        self.time = measured_value
+        self.test = test
+        self.original_verdict = original_verdict
+
+        test_name_max_length = max(test_name_max_length, len(self.test))
+
+    def to_str(
+        self,
+        limit: float,
+        segments: int,
+    ) -> str:
+        def get_bar() -> str:
+            percentage = self.time / limit
+            full = max(ceil(segments * percentage), 1)
+
+            bounded = min(full, segments)
+            overflown = max(0, full - segments)
+
+            overflown_max_length = terminal_width - segments - test_name_max_length - 40
+            if cut := (overflown > overflown_max_length):
+                overflown = overflown_max_length
+
+            if abs(percentage - 1) < 0.1:
+                color = "red"
+            elif abs(percentage - 1) < 0.5:
+                color = "yellow"
+            else:
+                color = "green"
+
+            full_bar = colored("━", color)
+
+            return (
+                f"[{full_bar*bounded}{'━'*(segments-bounded)}|"
+                f"{full_bar*overflown}{' '*(overflown_max_length - overflown)}{'⋯⋯' if cut else '  '}"
+            )
 
-def in_subtask(name: str, subtask: SubtaskConfig):
-    return any(fnmatch.fnmatch(f"{name}.in", g) for g in subtask.all_globs)
+        return (
+            f"{pad(self.test, test_name_max_length)}  "
+            f"{self.original_verdict.mark()}->{self.verdict.mark()}  "
+            f"{get_bar()}    {self.time:.2f} / {limit:.2f}s"
+        )
 
 
-def evaluate_solution(
-    results: dict[int, list[TestCaseResult]], config: TaskConfig
-) -> float:
-    points = 0.0
-    for subtask_id, sub_results in results.items():
-        points += evaluate_subtask(sub_results, config.subtasks[subtask_id].points)
-    return points
+def limit_result(result: LoggedResult, limit: float) -> LoggedResult:
+    if result.time <= limit and result.verdict == Verdict.timeout:
+        new_verdict = result.original_verdict
+        if new_verdict == Verdict.timeout:
+            new_verdict = Verdict.ok
 
+        return LoggedResult(
+            new_verdict, 1.0, result.time, result.test, result.original_verdict
+        )
+    if result.time > limit:
+        return LoggedResult(
+            Verdict.timeout, 0.0, result.time, result.test, result.original_verdict
+        )
```

Comment:
This method will always set `points` to either `0.0` or `1.0`. Is this desired?

Reply:
Not really. If neither case is true, we return original result. We could keep the original points, but this isn't consistent with new results.

We could not touch the points if you think that is better.

Reply:
Sorry, I missed the final `return`. This way makes sense.